Framework/Beamline/src/DetectorInfo.cpp

```diff
@@ -78,28 +90,236 @@ bool DetectorInfo::isEquivalent(const DetectorInfo &other) const {
   return true;
 }
 
-/// Returns the size of the DetectorInfo, i.e., the number of detectors in the
-/// instrument.
+/** Returns the number of detectors in the instrument.
+ *
+ * If a detector is moving, i.e., has more than one associated position, it is
+ * nevertheless only counted as a single detector. */
 size_t DetectorInfo::size() const {
-  if (!m_isMasked)
+  if (!m_isMonitor)
     return 0;
-  return m_isMasked->size();
+  return m_isMonitor->size();
 }
 
-/// Returns true if the detector is a monitor.
+/// Returns true if the beamline has scanning detectors.
+bool DetectorInfo::isScanning() const { return size() != m_positions->size(); }
+
+/// Returns true if the detector with given detector index is a monitor.
 bool DetectorInfo::isMonitor(const size_t index) const {
+  // No check for time dependence since monitor flags are not time dependent.
   return (*m_isMonitor)[index];
 }
 
-/// Returns true if the detector is masked.
+/** Returns true if the detector with given index is a monitor.
+ *
+ * The time component of the index is ignored since a detector is a monitor
+ * either for *all* times or for *none*. */
+bool DetectorInfo::isMonitor(const std::pair<size_t, size_t> index) const {
+  // Monitors are not time dependent, ignore time component of index.
+  return (*m_isMonitor)[index.first];
+}
+
+/** Returns true if the detector with given detector index is masked.
+ *
+ * Convenience method for beamlines with static (non-moving) detectors.
+ * Throws if there are time-dependent detectors. */
 bool DetectorInfo::isMasked(const size_t index) const {
+  checkNoTimeDependence();
   return (*m_isMasked)[index];
 }
 
-/// Set the mask flag of the detector with given index. Not thread safe.
+/// Returns true if the detector with given index is masked.
+bool DetectorInfo::isMasked(const std::pair<size_t, size_t> index) const {
+  return (*m_isMasked)[linearIndex(index)];
+}
+
+/** Set the mask flag of the detector with given detector index. Not thread
+ * safe.
+ *
+ * Convenience method for beamlines with static (non-moving) detectors.
+ * Throws if there are time-dependent detectors. */
 void DetectorInfo::setMasked(const size_t index, bool masked) {
+  checkNoTimeDependence();
   m_isMasked.access()[index] = masked;
 }
 
+/// Set the mask flag of the detector with given index. Not thread safe.
+void DetectorInfo::setMasked(const std::pair<size_t, size_t> index,
+                             bool masked) {
+  m_isMasked.access()[linearIndex(index)] = masked;
+}
+
+/// Returns the position of the detector with given index.
+Eigen::Vector3d
+DetectorInfo::position(const std::pair<size_t, size_t> index) const {
```

Comment:
Yes, it allows for generic code:

```cpp
for(const auto &index : spectrumInfo.spectrumDefinition(i))
  pos + =detectorInfo.position(index);
```

The time index would always be zero in the non-scanning case, but there is no need for client code to know that.